react_people-table-basics

DemetriodosAnjos opened the pull request:
[DEMO LINK - https://DemetriodosAnjos.github.io/react_people-table-basics/

DemetriodosAnjos committed:
fix: remove unused vars and type tests; fix PeopleTable render

mateacademy-ai-mentor commented on src/App.tsx:
```diff
@@ -1,167 +1,76 @@
-import { Loader } from './components/Loader';
+// Bloco Imports - reúne todas as importações
+import React from 'react';
+import { Routes, Route, Navigate, NavLink } from 'react-router-dom';
```
The task requires the app to use HashRouter so URLs are shareable on GitHub Pages. This file imports react-router-dom components but does not (and should not necessarily) instantiate HashRouter here — the HashRouter must be used at the app entry (index.tsx). I did not find the entry file to confirm HashRouter is applied, so please add or confirm an index.tsx that wraps <App /> with <HashRouter>. Without it, hash-based links (PersonLink's href="#/people/...") will not behave as intended. Reference: task description requires HashRouter.

mateacademy-ai-mentor replied:
App imports routing elements from react-router-dom — good. IMPORTANT: the task requires using HashRouter so app URLs are shareable on GitHub Pages. I don't see an entry file here wrapping `<App />` with `<HashRouter>`. Please ensure `src/index.tsx` (the app entry) wraps the app with `HashRouter` (imported from 'react-router-dom') to satisfy the requirement in the description.

mateacademy-ai-mentor commented on src/App.tsx:
```diff
@@ -1,167 +1,76 @@
+            <Route path="/" element={<HomePage />} />
+            <Route path="/people" element={<PeoplePage />} />
+            <Route path="/people/:slug" element={<PeoplePage />} />
+            <Route path="/home" element={<Navigate to="/" replace />} />
```
Good: Redirect from "/home" uses <Navigate to="/" replace /> which satisfies the requirement to avoid adding /home to history and prevent navigation loops. Keep this as-is.

mateacademy-ai-mentor replied:
Good: redirect from `/home` to `/` is implemented using `<Navigate to="/" replace />`, which uses the `replace` attribute as required to avoid saving `/home` in browser history and prevent navigation loops.

mateacademy-ai-mentor commented on src/components/PeoplePage.tsx:
```diff
@@ -0,0 +1,53 @@
+
+// Bloco PeoplePage - componente da página usando usePeople e Retry
+const PeoplePage: React.FC = () => {
+  const { slug } = useParams<{ slug?: string }>();
+  const { people, loading, error, reload } = usePeople();
+
+  return (
+    <div>
+      <h1 className="title">People Page</h1>
+
+      <div className="block">
+        {loading && (
+          <div className="box table-container">
+            <Loader />
+          </div>
+        )}
+
+        {!loading && error && (
+          <div>
+            <p data-cy="peopleLoadingError" className="has-text-danger">
+              {error}
+            </p>
+            <button
+              data-cy="retryButton"
+              className="button is-danger"
+              onClick={reload}
+              type="button"
+            >
+              Retry
+            </button>
+          </div>
+        )}
+
+        {!loading && !error && people.length === 0 && (
+          <p data-cy="noPeopleMessage">There are no people on the server</p>
+        )}
+
+        {!loading && !error && people.length > 0 && (
+          <PeopleTable people={people} selectedSlug={slug} />
```
Good: You correctly read the slug route param and pass it into PeopleTable as selectedSlug. This satisfies the selection/highlighting requirement (clicking or navigating to /people/:slug will result in the selected row highlighted by PeopleTable). See useParams + passing slug to <PeopleTable />.

mateacademy-ai-mentor replied:
Good: the component correctly renders all four UI states — loading (Loader), error (message + Retry), empty (noPeopleMessage), and data (PeopleTable). It also correctly reads `slug` via useParams and passes it to PeopleTable as `selectedSlug`, which keeps the selection logic inside the table as required  .

mateacademy-ai-mentor commented on src/components/PeoplePage.tsx:
```diff
@@ -0,0 +1,53 @@
+        )}
+
+        {!loading && !error && people.length === 0 && (
+          <p data-cy="noPeopleMessage">There are no people on the server</p>
```
Good: Loading / error / empty list states are handled explicitly: the Loader is shown during loading, an error message + retry button are shown on error (retry calls reload from usePeople), and a no-people message is displayed when the array is empty. This aligns with data lifecycle requirements.

mateacademy-ai-mentor commented on src/components/PeoplePage.tsx:
```diff
@@ -0,0 +1,53 @@
+  );
+};
+
+export default PeoplePage;
```
Reminder / action item (project-level): PersonLink produces hash URLs (e.g. '#/people/:slug'). Ensure the application root is wrapped with HashRouter in your entry file (index.tsx / main.tsx). The PeoplePage implementation relies on router param matching to work correctly when links are used or URL visited directly. If you share the index file I can verify or suggest the exact wrapper code.

mateacademy-ai-mentor commented on src/components/PeoplePage.tsx:
```diff
@@ -0,0 +1,53 @@
+      <div className="block">
+        {loading && (
+          <div className="box table-container">
+            <Loader />
```
Suggestion: For more idiomatic router navigation, consider switching PersonLink from a raw anchor to react-router-dom's <Link to={`/people/${person.slug}`}> (then HashRouter will create hash URLs automatically). This keeps navigation consistent with NavLink and Navigate used elsewhere. This is optional but recommended for SPA behavior.

mateacademy-ai-mentor commented on src/components/PeopleTable/PeopleTable.tsx:
```diff
@@ -0,0 +1,85 @@
+      return undefined;
+    }
+
+    const target = normalize(name);
+
+    return people.find(p => normalize(p.name) === target);
```
`findByName` correctly normalizes input and returns undefined for missing names. Good defensive handling prevents errors when parent name is absent. (See findByName implementation).

mateacademy-ai-mentor replied:
Good: `findByName` uses normalization (trim + toLowerCase) and then performs a find by normalized name. That correctly implements case-insensitive and whitespace-tolerant name matching for resolving mother/father references in the dataset.

mateacademy-ai-mentor commented on src/components/PeopleTable/PeopleTable.tsx:
```diff
@@ -0,0 +1,85 @@
+              <tr
+                key={person.slug}
+                data-cy="person"
+                className={isSelected ? 'has-background-warning' : ''}
```
Selected-row highlighting is implemented correctly using `selectedSlug === person.slug` and applying `has-background-warning`. This satisfies the row-selection requirement.

mateacademy-ai-mentor commented on src/components/PeopleTable/PeopleTable.tsx:
```diff
@@ -0,0 +1,85 @@
+                className={isSelected ? 'has-background-warning' : ''}
+              >
+                <td data-cy="personName">
+                  <PersonLink person={person} />
```
Rendering of the main person name uses `<PersonLink person={person} />` which is correct. Make sure `PersonLink` produces router-compatible links (or the app is wrapped with `HashRouter`) so navigation behaves as intended. If `PersonLink` is a raw anchor, the app must use HashRouter or you should switch to react-router's `Link`. See PersonLink for current implementation.

mateacademy-ai-mentor replied:
Good: main name column uses `PersonLink` so the person name is rendered as an internal link to `/people/:slug` for existing people. This matches the requirement that the main name be a link for existing persons and leverages the reusable component for consistent behavior.

mateacademy-ai-mentor commented on src/components/PeopleTable/PeopleTable.tsx:
```diff
@@ -0,0 +1,85 @@
+                  {mother ? (
+                    <PersonLink person={mother} />
+                  ) : (
+                    person.motherName || '-'
```
Mother cell: logic falls back to plain text or `-` when parent not found, which matches requirements. Note: you add `has-text-danger` on the `<td>` when `mother?.sex === 'f'` while `PersonLink` also applies `has-text-danger` for female names. This duplicates the styling; consider applying the class in only one place (preferably `PersonLink`) for clarity.

mateacademy-ai-mentor commented on src/components/PeopleTable/PeopleTable.tsx:
```diff
@@ -0,0 +1,85 @@
+                  {father ? (
+                    <PersonLink person={father} />
+                  ) : (
+                    person.fatherName || '-'
```
Father cell behavior mirrors the mother handling and is correct: render `PersonLink` when found, otherwise `person.fatherName || '-'`. Good.

mateacademy-ai-mentor commented on src/components/PersonLink/PersonLink.tsx:
```diff
@@ -0,0 +1,20 @@
+  return (
+    <a className={className} href={`#/people/${person.slug}`}>
+      {person.name}
+    </a>
```
The anchor uses a hash-based href: `<a className={className} href={`#/people/${person.slug}`}>`. This will work only if your app root is wrapped with `HashRouter`. For cleaner, SPA-friendly navigation consider replacing the anchor with `Link` from react-router-dom: `<Link className={className} to={`/people/${person.slug}`}>{person.name}</Link>`. If you keep the anchor, ensure index.tsx wraps the app with `<HashRouter>` so hash navigation works as the task requires. See this file for the anchor usage.

mateacademy-ai-mentor commented on src/components/PersonLink/PersonLink.tsx:
```diff
@@ -0,0 +1,20 @@
+// Bloco PersonLink - componente reaproveitável
+export const PersonLink: React.FC<{ person?: Person }> = ({ person }) => {
+  if (!person) {
+    return <span>-</span>;
```
When `person` is falsy you return `<span>-</span>`. This is acceptable because PeopleTable only calls PersonLink for existing person objects, but consider returning `null` or a plain text label instead if you plan to reuse PersonLink in other contexts (so behavior is explicit and consistent).

mateacademy-ai-mentor commented on src/components/PersonLink/PersonLink.tsx:
```diff
@@ -0,0 +1,20 @@
+    return <span>-</span>;
+  }
+
+  const className = person.sex === 'f' ? 'has-text-danger' : '';
```
The className calculation is correct and satisfies the requirement to apply `has-text-danger` for women (`person.sex === 'f'`). Good — this ensures female names get the proper styling.

mateacademy-ai-mentor commented on src/hooks/__test__/usePeople.test.tsx:
```diff
@@ -0,0 +1,53 @@
+import { renderHook } from '@testing-library/react-hooks';
+import { usePeople } from '../usePeople';
```
The import on this line imports `usePeople` as a named export (`{ usePeople }`) but the hook file exports the hook as the default export. This mismatch will make `usePeople` undefined in the test. Either change this to a default import `import usePeople from '../usePeople'` or change the hook file to export a named function. Reference: src/hooks/usePeople.ts exports default.

mateacademy-ai-mentor commented on src/hooks/__test__/usePeople.test.tsx:
```diff
@@ -0,0 +1,53 @@
+  });
+
+  it('loads people', async () => {
+    const { result, waitForNextUpdate } = renderHook(() => usePeople());
```
Because the imported symbol will be undefined given the current import (line 2), the renderHook call here will fail. After you fix the import/export mismatch, this line is fine as-is: `renderHook(() => usePeople())` properly tests the hook and waits for updates. If you switch to default import, no other changes needed here.

mateacademy-ai-mentor commented on src/hooks/usePeople.ts:
```diff
@@ -0,0 +1,40 @@
+  'https://mate-academy.github.io/react_people-table/api/people.json';
+
+// Bloco usePeople - hook que busca pessoas e expõe reload
+export default function usePeople() {
```
The hook is exported as default here. The tests import `usePeople` as a named export (see src/hooks/__test__/usePeople.test.tsx). This mismatch will cause test failures. Consider adding a named export or changing the test to a default import. Example options:
- add `export function usePeople() { ... }` and `export default usePeople` OR
- keep default export and change the test to `import usePeople from '../usePeople'`.
Reference: test file currently does `import { usePeople } from '../usePeople'`.

mateacademy-ai-mentor commented on src/hooks/usePeople.ts:
```diff
@@ -0,0 +1,40 @@
+      const data: Person[] = await res.json();
+
+      setPeople(data);
+    } catch {
```
The catch block currently swallows the error (no parameter). For easier debugging, capture the error and log it (or include its message in the `error` state). Example: `} catch (err) { setPeople([]); setError('Something went wrong'); console.error(err); }`. This helps when CI/tests fail due to network/JSON issues.

mateacademy-ai-mentor commented on src/hooks/usePeople.ts:
```diff
@@ -0,0 +1,40 @@
+  }, []);
+
+  useEffect(() => {
+    fetchPeople();
```
Consider adding a cancellation mechanism (AbortController) or a mounted flag to avoid calling setState after unmount when a fetch is still pending. Not required by the task/tests but a good robustness improvement for real apps. Keep the current implementation for simplicity if you prefer.

mateacademy-ai-mentor commented on src/hooks/usePeople.ts:
```diff
@@ -0,0 +1,40 @@
+    fetchPeople();
+  }, [fetchPeople]);
+
+  return { people, loading, error, reload: fetchPeople };
```
Good: fetchPeople is memoized with useCallback and is used as a dependency in useEffect, so the hook fetches once on mount and exposes reload. Returning `reload: fetchPeople` is explicit and convenient for the page to retry. No change needed.

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

I'm requesting changes because there are two critical blockers that can break functionality and/or tests:

- HashRouter wrapper missing/unverified (critical requirement): The task explicitly requires HashRouter so links work on GitHub Pages. I couldn't find the entry file (e.g., src/index.tsx) to confirm this. Without wrapping `<App />` in `<HashRouter>`, hash-based links like `#/people/:slug` won’t behave correctly when shared or directly opened. Fix: in `src/index.tsx`, wrap the app with `HashRouter`:
  - import `{ HashRouter }` from `react-router-dom`
  - render `<HashRouter><App /></HashRouter>`

- usePeople export/import mismatch in tests (will fail tests): `src/hooks/usePeople.ts` exports the hook as default, but `src/hooks/__test__/usePeople.test.tsx` imports `{ usePeople }` as a named export. This makes `usePeople` undefined and the test will fail. Fix one of:
  - Change the test import to default: `import usePeople from '../usePeople'`, or
  - Add a named export in the hook and also keep default:
    - `export function usePeople() { ... }`
    - `export default usePeople`

What’s good: routing and pages in App.tsx are correct (Home, People, /home redirect with `replace`, NotFound), PeoplePage handles loading/error/empty states and slug selection well, PeopleTable renders and highlights correctly, and PersonLink applies female styling.

Minor (non-blocking) suggestions:
- Consider switching `PersonLink` to `Link` from `react-router-dom` (`to={`/people/${person.slug}`}`) for idiomatic SPA navigation.
- Avoid duplicate `has-text-danger` styling on the mother cell and `PersonLink`.
- Optionally log the caught error in `usePeople` for easier debugging.

Address the two blockers above, and you should be set. Nice work on the core routing and table logic!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

DemetriodosAnjos committed:
fix(hooks): add eslint comment for console.error to satisfy no-console rule

mateacademy-ai-mentor commented on src/App.tsx:
```diff
@@ -1,167 +1,76 @@
+      >
+        <div className="container">
+          <div className="navbar-brand">
+            <NavLink
+              to="/"
+              end
```
Good: the Home NavLink uses the `end` prop and a className callback that applies `has-background-grey-lighter` when active. This satisfies the 'active link should have has-background-grey-lighter' requirement and prevents Home from being active on child routes.

mateacademy-ai-mentor commented on src/components/PeoplePage.tsx:
```diff
@@ -0,0 +1,53 @@
+// Bloco PeoplePage - componente da página usando usePeople e Retry
+const PeoplePage: React.FC = () => {
+  const { slug } = useParams<{ slug?: string }>();
+  const { people, loading, error, reload } = usePeople();
```
PeoplePage destructures `reload` from `usePeople()` but the hook currently does not return a `reload` function. The Retry button depends on this; please add and return `reload` (or rename consistently) from the hook so the page can call it without error. See the hook implementation to update its return shape and behavior .

mateacademy-ai-mentor commented on src/components/PeoplePage.tsx:
```diff
@@ -0,0 +1,53 @@
+            <p data-cy="peopleLoadingError" className="has-text-danger">
+              {error}
+            </p>
+            <button
+              data-cy="retryButton"
+              className="button is-danger"
+              onClick={reload}
+              type="button"
```
The Retry button's onClick uses `reload`. If the hook doesn't return `reload`, clicking Retry will be a no-op or throw. Implement `reload` in the hook (and ensure it restarts the fetch safely, using AbortController or equivalent) so this handler works as intended. Also ensure the hook's fetch uses the required API endpoint (see hook file) .

mateacademy-ai-mentor commented on src/components/PeopleTable/PeopleTable.tsx:
```diff
@@ -0,0 +1,85 @@
+
+                <td
+                  data-cy="personMother"
+                  className={mother?.sex === 'f' ? 'has-text-danger' : ''}
```
Minor issue: you apply `has-text-danger` both on the mother `<td>` (here) and inside `PersonLink` for female persons. This duplicates the class on nested elements. Consider removing the `className` on the `<td>` and keeping the styling inside `PersonLink` (or vice versa) to avoid redundancy and keep styling logic in one place.

mateacademy-ai-mentor commented on src/components/PersonLink/PersonLink.tsx:
```diff
@@ -0,0 +1,19 @@
+import { Link } from 'react-router-dom';
+import { Person } from '../../types/Person';
+
+export const PersonLink: React.FC<{ person?: Person }> = ({ person }) => {
+  if (!person) {
+    return <span>-</span>;
+  }
+
+  const className = person.sex === 'f' ? 'has-text-danger' : '';
+
+  return (
+    <Link className={className} to={`/people/${person.slug}`}>
```
Good: The component matches the requested API `<PersonLink person={person} />` and returns a `Link` to `/people/:slug`. It also applies `has-text-danger` when `person.sex === 'f'`, which satisfies the requirement to highlight women’s names. Nice and concise implementation.

mateacademy-ai-mentor commented on src/components/PersonLink/PersonLink.tsx:
```diff
@@ -0,0 +1,19 @@
+import { Person } from '../../types/Person';
+
+export const PersonLink: React.FC<{ person?: Person }> = ({ person }) => {
+  if (!person) {
+    return <span>-</span>;
```
When `person` is falsy you return `<span>-</span>`. This is harmless because PeopleTable renders `person.motherName || '-'` or `person.fatherName || '-'` itself, but consider returning `null` instead to keep responsibilities separated — let the table decide how to render missing parents. This change is optional but recommended for clearer component semantics.

mateacademy-ai-mentor commented on src/components/PersonLink/PersonLink.tsx:
```diff
@@ -0,0 +1,19 @@
+import React from 'react';
+import { Link } from 'react-router-dom';
```
Using `Link` from `react-router-dom` is correct for SPA navigation. Make sure the application entrypoint wraps the app with `HashRouter` so these links work as expected on GitHub Pages (the component assumes a router context is present).

mateacademy-ai-mentor commented on src/components/PersonLink/PersonLink.tsx:
```diff
@@ -0,0 +1,19 @@
+      {person.name}
+    </Link>
+  );
+};
+
+export default PersonLink;
```
Note: you export both a named export `PersonLink` and a default export. That's fine, but be consistent where you import this component across the codebase/tests (named vs default) to avoid import mismatches.

mateacademy-ai-mentor commented on src/hooks/usePeople.ts:
```diff
@@ -0,0 +1,67 @@
+import { useEffect, useState } from 'react';
+import { Person } from '../types/Person';
+
+const API_URL = '/api/people';
```
The API URL constant is set to `'/api/people'` but the task requires fetching from `https://mate-academy.github.io/react_people-table/api/people.json`. Update `API_URL` to the required endpoint or make it configurable with that URL as the default so the hook fetches the correct dataset (checklist items about fetching people from the given API) .

mateacademy-ai-mentor commented on src/hooks/usePeople.ts:
```diff
@@ -0,0 +1,67 @@
+    const controller = new AbortController();
+    const { signal } = controller;
+
+    const load = async () => {
```
`load` is declared inside the effect. To support exposing a `reload` function you need `load` to be callable from outside the effect (or use a reload counter/state that the effect depends on). Consider moving `load` out of the effect and wrapping it in `useCallback`, using a `useRef` to hold `load`, or implementing a `reloadCounter` state and returning a `reload` that increments it. Also ensure AbortController lifecycle is handled correctly for each invocation.

mateacademy-ai-mentor commented on src/hooks/usePeople.ts:
```diff
@@ -0,0 +1,67 @@
+    };
+  }, []);
+
+  return { people, loading, error };
```
The hook currently returns `{ people, loading, error }`, but `PeoplePage` expects `reload` and calls it from the Retry button (`onClick={reload}`). Add `reload` to the returned object (for example `return { people, loading, error, reload }`) so the consumer can re-trigger a fetch. See PeoplePage which destructures `reload` from the hook and uses it — the page will fail otherwise .

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

Thanks for the solid work — routing, the /home redirect with replace, NavLink active styling, PeopleTable behavior, and PersonLink usage are all implemented well. However, there are a few critical blockers that need fixing to meet the task requirements and avoid runtime issues:

- Critical: HashRouter not confirmed in entry point (src/index.tsx). The task requires using HashRouter for GitHub Pages. Without wrapping `<App />` in `<HashRouter>`, direct links won’t work. Fix: in `src/index.tsx`, import `HashRouter` from `react-router-dom` and render `<HashRouter><App /></HashRouter>`.
- Critical: Wrong API endpoint in `src/hooks/usePeople.ts` (line with `API_URL = '/api/people'`). Requirement is to fetch from `https://mate-academy.github.io/react_people-table/api/people.json`. Update the URL so PeoplePage loads the correct dataset.
- Critical: Missing `reload` in `usePeople` hook API. `PeoplePage.tsx` destructures `reload` and uses it for the Retry button, but the hook only returns `{ people, loading, error }`. This breaks Retry and can cause runtime issues. Fix: implement and return a `reload` function (e.g., via a version state trigger or a `useCallback`ed fetch function) and include it in the return: `{ people, loading, error, reload }`.

Minor suggestions (non-blockers):
- Remove duplicate `has-text-danger` class on the mother `<td>` in PeopleTable and keep female-name styling within `PersonLink` to avoid redundancy.
- Consider returning `null` instead of `'-'` from `PersonLink` when `person` is falsy, letting the table handle the placeholder.

Once you add HashRouter, correct the API URL, and expose `reload`, your app should fully satisfy the requirements. Good job on the rest — you’re very close!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

DemetriodosAnjos committed:
fix(people): stabilize selectors, fix loading/error states, links
- make people null until loaded; set [] on success, null on error
- show loader, error and empty states reliably; ignore aborts
- move data-cy=person to <tr>, add per-cell selectors
- render parent names as links when present; female names red
- highlight selected row with has-background-warning

mateacademy-ai-mentor commented on src/App.scss:
```diff
@@ -1,3 +1,29 @@
+/* garantir que a tabela do componente seja o escopo */
+.table-container {
+  /* linha selecionada: amarelo suave */
+  table tr.is-selected {
+    background-color: #fff8c2 !important;
+  }
+
+  /* fallback para hover/seleção caso Bulma aplique variantes */
+  table tr.is-selected td {
+    background-color: #fff8c2 !important;
+  }
```
The task requires that the selected person’s row use the class `has-background-warning` (requirements 1.13 / 1.14). This stylesheet only targets `table tr.is-selected` and `table tr.is-selected td`. If your components apply `has-background-warning` to the selected row (as the spec requires), they may not receive this styling. Consider adding the same rules for `.has-background-warning`, for example:

- `table tr.has-background-warning` and
- `table tr.has-background-warning td`

so the required class is styled identically to `.is-selected`. This ensures the app meets the checklist exactly and avoids depending on Bulma defaults.

mateacademy-ai-mentor commented on src/App.scss:
```diff
@@ -1,3 +1,29 @@
+    background-color: #fff8c2 !important;
+  }
+
+  /* cores explícitas para links de pessoa, com maior especificidade */
+  a.person-link--female.has-text-danger,
+  a.person-link--female {
+    color: #d9534f !important; /* vermelho */
+  }
```
You style `a.person-link--female.has-text-danger` and `a.person-link--female`. The task explicitly requires women’s names to have the class `has-text-danger` (requirement 1.12). Two notes:

- Prefer relying on `has-text-danger` for female name styling (it directly reflects the requirement and Bulma semantic class).
- If you keep `person-link--female`, ensure the `PersonLink` component actually renders both classes (`person-link--female` and `has-text-danger`) when appropriate, otherwise the `has-text-danger` requirement may not be met.

Consider simplifying to style `.has-text-danger` under `.table-container` or ensure the component applies both classes consistently.

mateacademy-ai-mentor commented on src/components/PeopleTable/PeopleTable.scss:
```diff
@@ -0,0 +1,31 @@
+/* Escopo do componente */
+.table-container {
+  /* linha selecionada: amarelo suave (maior especificidade para sobrepor regras conflitantes) */
+  table tr.is-selected,
+  table tr.is-selected td {
+    background-color: #fff8c2 !important;
+    color: #000 !important;
+  }
```
Critical: The selected-row selectors target `tr.is-selected`, but the task requires that the selected row be highlighted via the class `has-background-warning` (checklist items 1.13 and 1.14). Update the selectors to include `tr.has-background-warning` and the corresponding `td` selectors so the required class is actually styled. Example: `table tr.is-selected, table tr.has-background-warning, table tr.is-selected td, table tr.has-background-warning td { ... }`

mateacademy-ai-mentor commented on src/components/PeopleTable/PeopleTable.scss:
```diff
@@ -0,0 +1,31 @@
+    color: #000 !important;
+  }
+
+  /* Garantir contraste ao passar o mouse em linhas selecionadas */
+  table tr.is-selected:hover,
+  table tr.is-selected:hover td {
+    background-color: #fff6b8 !important;
+  }
```
Also include hover styling for the required class. The hover rule currently targets `table tr.is-selected:hover`; add `table tr.has-background-warning:hover` as well so the `has-background-warning` class gets the same hover behavior.

mateacademy-ai-mentor commented on src/components/PeopleTable/PeopleTable.scss:
```diff
@@ -0,0 +1,31 @@
+  /* linha selecionada: amarelo suave (maior especificidade para sobrepor regras conflitantes) */
+  table tr.is-selected,
+  table tr.is-selected td {
+    background-color: #fff8c2 !important;
+    color: #000 !important;
+  }
+
+  /* Garantir contraste ao passar o mouse em linhas selecionadas */
+  table tr.is-selected:hover,
+  table tr.is-selected:hover td {
+    background-color: #fff6b8 !important;
+  }
+
+  /* Cores explícitas para links de pessoa (especificidade + !important para anular overrides) */
+  a.person-link--female,
+  a.person-link--female.has-text-danger {
+    color: #d9534f !important; /* vermelho */
+  }
+
+  a.person-link--male,
+  a.person-link--male.has-text-primary {
+    color: #3273dc !important; /* azul */
```
The file defines styles for `a.person-link--female` and `a.person-link--male`, but the task mandates that women’s names must carry the class `has-text-danger` (requirement 1.12). To ensure the required class results in the expected color, add selectors for `a.has-text-danger` (and `a.has-text-primary` if you rely on that) or use `a.has-text-danger` as the primary selector for female links. This avoids coupling visual outcome to a custom class that components may not set.

mateacademy-ai-mentor replied:
There are multiple `!important` uses (e.g., background and color rules at lines 8,9,15,21,26). Excessive `!important` can mask whether the application code is applying the correct classes (e.g., `has-text-danger`, `has-background-warning`). Prefer using more specific selectors scoped to `.table-container` rather than `!important` so required classes remain meaningful and debuggable.

mateacademy-ai-mentor commented on src/components/PeopleTable/PeopleTable.scss:
```diff
@@ -0,0 +1,31 @@
+/* src/components/PeopleTable/PeopleTable.scss */
+
+/* Escopo do componente */
+.table-container {
+  /* linha selecionada: amarelo suave (maior especificidade para sobrepor regras conflitantes) */
+  table tr.is-selected,
+  table tr.is-selected td {
+    background-color: #fff8c2 !important;
+    color: #000 !important;
+  }
+
+  /* Garantir contraste ao passar o mouse em linhas selecionadas */
+  table tr.is-selected:hover,
+  table tr.is-selected:hover td {
+    background-color: #fff6b8 !important;
+  }
+
+  /* Cores explícitas para links de pessoa (especificidade + !important para anular overrides) */
+  a.person-link--female,
+  a.person-link--female.has-text-danger {
+    color: #d9534f !important; /* vermelho */
+  }
+
+  a.person-link--male,
+  a.person-link--male.has-text-primary {
+    color: #3273dc !important; /* azul */
+  }
+
+  /* Se houver outras regras globais interferindo (ex: .has-text-primary alterada),
+     assegure-se que essas regras do componente vençam por especificidade. */
+}
```
Small suggestion: your scoping (`.table-container`) is good. Make sure the component actually applies the expected class names from the task (e.g., `has-background-warning` on the selected row and `has-text-danger` for female names) — the stylesheet should support those exact class names so the checklist requirements are satisfied without requiring additional custom classes.

mateacademy-ai-mentor commented on src/components/PeopleTable/PeopleTable.tsx:
```diff
@@ -0,0 +1,288 @@
+              </td>
+            </tr>
+          ) : (
+            sortedPeople.map(person => {
+              const isSelected = Boolean(
+                selectedSlug && person.slug === selectedSlug,
+              );
```
The expression used to compute `isSelected` is unnecessarily complex and uses `Boolean(...)` with a trailing comma. This is confusing and less readable. Prefer the simpler and equivalent form:

const isSelected = selectedSlug === person.slug;

This is clearer and handles undefined `selectedSlug` correctly (it will be false).

mateacademy-ai-mentor commented on src/components/PeopleTable/PeopleTable.tsx:
```diff
@@ -0,0 +1,288 @@
+                  key={person.slug}
+                  data-cy="person"
+                  data-cy-row={`person-row-${person.slug}`}
+                  className={isSelected ? 'has-background-warning' : undefined}
```
Note: you set `className={isSelected ? 'has-background-warning' : undefined}` which is fine. If you later need to apply multiple classes, use a helper or template literal so you don't overwrite other classes unintentionally (e.g., `className={isSelected ? 'has-background-warning other-class' : 'other-class'}`).

mateacademy-ai-mentor commented on src/components/PeopleTable/PeopleTable.tsx:
```diff
@@ -0,0 +1,288 @@
+
+    return base;
+  };
+
+  const findByName = (name?: string) =>
+    name ? people.find(p => p.name === name) : undefined;
```
Behavioral note: `findByName` uses strict equality `p.name === name` to find parent objects. This satisfies the requirement, but be aware this will not match names with different casing or extra whitespace. If the dataset may have inconsistencies, consider normalizing/triming both sides before comparison.

mateacademy-ai-mentor commented on src/components/PersonLink/PersonLink.tsx:
```diff
@@ -0,0 +1,26 @@
+import { Person } from '../../types/Person';
+
+export const PersonLink: React.FC<{ person?: Person }> = ({ person }) => {
+  if (!person) {
+    return null;
```
Behavior issue: the component returns `null` when `person` is falsy. The task requires that PersonLink render plain text when no matching person exists (checklist item 3.1). Instead of returning `null`, consider rendering the fallback name as plain text or accept a `name`/`fallbackName` prop so callers can pass the parent name to display.

mateacademy-ai-mentor commented on src/components/PersonLink/PersonLink.tsx:
```diff
@@ -0,0 +1,26 @@
+  return (
+    <Link
+      className={`${sexClass} ${bulmaClass}`}
+      to={`/people/${person.slug}`}
```
Safety: the code uses `person.slug` directly to construct the `to` prop. If a `Person` object lacks a slug, this will produce an invalid link. Add a guard to ensure `person.slug` exists and fall back to rendering plain text when it's missing (requirement 3.1 also implies this case).

mateacademy-ai-mentor commented on src/components/PersonLink/PersonLink.tsx:
```diff
@@ -0,0 +1,26 @@
+
+  const sexClass =
+    person.sex === 'f' ? 'person-link--female' : 'person-link--male';
+  const bulmaClass =
+    person.sex === 'f' ? 'has-text-danger' : 'has-text-primary';
+
+  return (
+    <Link
+      className={`${sexClass} ${bulmaClass}`}
```
Styling note: you add both a custom class (`person-link--female`) and the Bulma class (`has-text-danger`) which is good. Ensure your component and stylesheet rely on the required `has-text-danger` class (checklist item 1.12) so the visual styling is enforced even if custom classes change.

mateacademy-ai-mentor commented on src/components/PersonLink/PersonLink.tsx:
```diff
@@ -0,0 +1,26 @@
+      {person.name}
+    </Link>
+  );
+};
+
+export default PersonLink;
```
Export style: you export the component twice — as a named export and as default. This is allowed, but be consistent across the codebase. If the project prefers one style, align to that. No functional blocker here, just a consistency remark.

mateacademy-ai-mentor commented on src/hooks/__test__/usePeople.test.tsx:
```diff
@@ -0,0 +1,53 @@
+type FetchMock = jest.Mock<Promise<Response>, [RequestInfo, RequestInit?]>;
+
+describe('usePeople', () => {
+  const originalFetch = global.fetch;
```
Store/restore of `global.fetch` may be fragile if `global.fetch` is undefined in the test environment. Consider capturing it with a typed variable or using `jest.spyOn(global, 'fetch')` for safer mocking and restoration. This will make tear-down more robust across environments.

mateacademy-ai-mentor commented on src/hooks/__test__/usePeople.test.tsx:
```diff
@@ -0,0 +1,53 @@
+  beforeEach(() => {
+    const mockJson = () =>
+      Promise.resolve(mockPeople) as Promise<PersonResponse[]>;
+    const mockResponse = {
+      ok: true,
+      json: mockJson,
```
The `mockResponse`'s `json` implementation returning `mockPeople` is fine, but the test should assert that the hook called `fetch` with the exact required API URL `https://mate-academy.github.io/react_people-table/api/people.json`. Add an expectation such as `expect(global.fetch).toHaveBeenCalledWith('https://mate-academy.github.io/react_people-table/api/people.json')` after awaiting the update to ensure the correct endpoint is used (checklist items 1.7 and 2.3).

mateacademy-ai-mentor commented on src/hooks/__test__/usePeople.test.tsx:
```diff
@@ -0,0 +1,53 @@
+      json: mockJson,
+    } as unknown as Response;
+
+    global.fetch = jest.fn(() => Promise.resolve(mockResponse)) as FetchMock;
```
Instead of directly assigning `global.fetch = jest.fn(...)`, consider `jest.spyOn(global, 'fetch').mockResolvedValue(mockResponse as unknown as Response)`. This approach is cleaner and makes it easier to restore the original implementation with `mockRestore()` in `afterEach`.

mateacademy-ai-mentor commented on src/hooks/__test__/usePeople.test.tsx:
```diff
@@ -0,0 +1,53 @@
+    const { result, waitForNextUpdate } = renderHook(() => usePeople());
+
+    await waitForNextUpdate();
+    expect(result.current.people.length).toBeGreaterThan(0);
```
The test currently asserts only that `people.length` is greater than 0. Strengthen the test by checking the full hook contract: e.g., `expect(result.current.loading).toBe(false)`, `expect(result.current.error).toBeNull()` (or falsy), and `expect(typeof result.current.reload).toBe('function')`. If `reload` exists, also consider invoking it and asserting that `fetch` is called again. This protects against regressions where `reload` is missing (previous review highlighted this issue).

mateacademy-ai-mentor commented on src/hooks/__test__/usePeople.test.tsx:
```diff
@@ -0,0 +1,53 @@
+  it('loads people', async () => {
+    const { result, waitForNextUpdate } = renderHook(() => usePeople());
+
+    await waitForNextUpdate();
```
`waitForNextUpdate` is acceptable with older `@testing-library/react-hooks` versions, but if you upgrade, prefer `waitFor` from `@testing-library/react` which can be clearer and less tied to internal hook update mechanics. Consider switching if you encounter flakiness.

mateacademy-ai-mentor commented on src/hooks/usePeople.ts:
```diff
@@ -0,0 +1,78 @@
+import { useCallback, useEffect, useState } from 'react';
+import { Person } from '../types/Person';
+
+const API_URL =
+  'https://mate-academy.github.io/react_people-table/api/people.json';
```
Good: the API URL is the exact required endpoint for the task, so the hook will fetch the correct dataset. This satisfies requirement 1.7 / 2.3.

mateacademy-ai-mentor commented on src/hooks/usePeople.ts:
```diff
@@ -0,0 +1,78 @@
+    };
+  }, [load, reloadIndex]);
+
+  const reload = useCallback(() => {
+    setReloadIndex(i => i + 1);
+  }, []);
```
Good: a `reload` function is implemented via `reloadIndex` and returned from the hook, so consumers can retry loading data. This satisfies the requirement that PeoplePage can call `reload` for the Retry action.

mateacademy-ai-mentor commented on src/hooks/usePeople.ts:
```diff
@@ -0,0 +1,78 @@
+      // success: even if array is empty, consider this a successful load
+      setPeople(Array.isArray(data) ? data : []);
+      setError(null);
+    } catch (err: unknown) {
+      // If the request was aborted, do not show a loading error in the UI
+      if (err instanceof Error && err.name === 'AbortError') {
+        // keep previous error (should be null) and just stop loading
```
Suggestion (robustness): the abort detection uses `err instanceof Error && err.name === 'AbortError'` (lines 38–41). In some environments the thrown abort error can be a `DOMException` or another object that still has `name === 'AbortError'` but may not pass the `instanceof Error` check. Consider checking the `name` property first (e.g., `if ((err as any)?.name === 'AbortError')`) or include `DOMException` in the check to be more robust across environments.

mateacademy-ai-mentor commented on src/hooks/usePeople.ts:
```diff
@@ -0,0 +1,78 @@
+    } finally {
+      setLoading(false);
+    }
+  }, []);
```
Suggestion (clarity): `load` is memoized with `useCallback(..., [])` (line 56). This is fine because `API_URL` is a module-level constant and state setters are stable, but it would help future readers to add a short comment explaining that the empty dependency array is intentional to keep the callback stable for the `useEffect` dependency.

mateacademy-ai-mentor commented on src/index.tsx:
```diff
@@ -1,12 +1,16 @@
+import React from 'react';
```
Small note: in modern React with the automatic JSX runtime the `import React from 'react'` import is usually not required. It's harmless to keep, but you can remove it if your build/config uses the new JSX transform.

mateacademy-ai-mentor commented on src/index.tsx:
```diff
@@ -1,12 +1,16 @@
 import '@fortawesome/fontawesome-free/css/all.css';
 
-import { App } from './App';
+import App from './App';
```
Good: App is imported from './App' — ensure `App` defines routes and that HashRouter usage here will satisfy the routing requirements for GitHub Pages (this file correctly provides HashRouter at the app root).

mateacademy-ai-mentor commented on src/index.tsx:
```diff
@@ -1,12 +1,16 @@
+import App from './App';
 
-createRoot(document.getElementById('root') as HTMLDivElement).render(
+const container = document.getElementById('root') as HTMLDivElement;
+const root = createRoot(container);
```
Runtime-safety: `document.getElementById('root')` is cast to `HTMLDivElement` without a runtime null check. If the element is not present, `createRoot(container)` will receive `null` and crash. Consider guarding explicitly, for example:

const container = document.getElementById('root');
if (!container) throw new Error('Root element not found');
const root = createRoot(container);

This makes the failure mode explicit and avoids silent type assertions that hide a possible runtime error.

mateacademy-ai-mentor commented on src/index.tsx:
```diff
@@ -1,12 +1,16 @@
+const container = document.getElementById('root') as HTMLDivElement;
+const root = createRoot(container);
+
+root.render(
```
Minor: createRoot and render usage is correct and the app is wrapped with HashRouter. This satisfies the requirement to use HashRouter so routes will work with hash-based URLs (good fix from previous review).

mateacademy-ai-mentor commented on src/index.tsx:
```diff
@@ -1,12 +1,16 @@
+root.render(
   <Router>
     <App />
   </Router>,
```
Stylistic: the JSX is wrapped inside `<Router><App /></Router>` which is exactly what's needed. No further changes required here unless you want to add StrictMode or other providers around App.

mateacademy-ai-mentor reviewed (approved):
### Review Summary

Educational approval: great progress overall. Routing and HashRouter setup look solid, the data hook fetches from the exact required URL and exposes reload, and PeopleTable implements sorting, correct columns, conditional parent links, and applies `has-background-warning` on selection. To align precisely with the checklist: update your styles to also target `tr.has-background-warning` (and its `td`, plus hover) alongside `tr.is-selected` so selected rows are styled when using the required class (items 1.13–1.14). Likewise, prefer styling `a.has-text-danger` for women’s names (item 1.12) or ensure `PersonLink` always renders both classes if you keep custom ones. In PeopleTable.tsx, simplify `isSelected` to `const isSelected = selectedSlug === person.slug;` for clarity. In PersonLink.tsx, instead of returning `null` when `person` is missing, render plain text and guard against missing `slug` to meet item 3.1. Tests: assert the exact fetch URL and that `reload` exists and works; consider `jest.spyOn(global, 'fetch')` for safer mocking. Nice work—these tweaks will make the app fully compliant and more maintainable.

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)
